### PR TITLE
Support parallel function calling 

### PR DESCRIPTION
**Pull request by jerryjliu**
- rename chat_with_tool to chat_with_tools 
- rename _get_tool_call_from_response to _get_tool_calls_from_response
- allow parallel function calling on predict_and_call  (by default false) 
- allow parallel function calling on FunctionCallingAgentWorker (by default true)

### CODE REVIEW

File: llama-index-integrations/llms/llama-index-llms-mistralai/llama_index/llms/mistralai/base.py [hunk 1]
```diff
@@ -372,34 +398,55 @@ async def apredict_and_call(
...
+            verbose=verbose,
+            allow_parallel_tool_calls=allow_parallel_tool_calls,
+            **kwargs,
         )
-        tool_call = self._get_tool_call_from_response(response)
-        tool_output = acall_tool_with_selection(tool_call, tools, verbose=verbose)
-        return AgentChatResponse(response=tool_output.content, sources=[tool_output])
+        tool_calls = self._get_tool_calls_from_response(response)
+        tool_outputs = [
+            acall_tool_with_selection(tool_call, tools, verbose=verbose)
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-mistralai/llama_index/llms/mistralai/base.py:428]: Should this use await? Should we use gather() as well?

File: llama-index-core/llama_index/core/agent/function_calling/step.py [hunk 2]
```diff
@@ -289,29 +305,43 @@ async def arun_step(
...
             is_done = False
-            await self._acall_function(
-                tools,
-                tool_call,
-                task.extra_state["new_memory"],
-                task.extra_state["sources"],
-                verbose=self._verbose,
-            )
+            for tool_call in tool_calls:
+                # TODO: maybe execute this with multi-threading
```
Comment (logan-markewich) [inline llama-index-core/llama_index/core/agent/function_calling/step.py:336]: Even gather could be better? (Most tools aren't async though, just fake async)

File: llama-index-integrations/llms/llama-index-llms-mistralai/llama_index/llms/mistralai/base.py (hunk 1 above)
Comment (jerryjliu) [reply llama-index-integrations/llms/llama-index-llms-mistralai/llama_index/llms/mistralai/base.py:428]: uhhhh oops

File: llama-index-integrations/llms/llama-index-llms-mistralai/llama_index/llms/mistralai/base.py [hunk 3]
```diff
@@ -297,12 +306,30 @@ def predict_and_call(
...
+            for tool_call in tool_calls
+        ]
+        if allow_parallel_tool_calls:
+            output_text = "\n\n".join(
+                [tool_output.content for tool_output in tool_outputs]
+            )
+            return AgentChatResponse(response=output_text, sources=tool_outputs)
+        else:
+            if len(tool_outputs) > 1:
+                raise ValueError("Invalid")
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-mistralai/llama_index/llms/mistralai/base.py:329]: Probably needs a better error msg

File: llama-index-integrations/llms/llama-index-llms-mistralai/llama_index/llms/mistralai/base.py [hunk 4]
```diff
@@ -372,34 +399,56 @@ async def apredict_and_call(
...
+        ]
+        tool_outputs = await asyncio.gather(*tool_tasks)
+        if allow_parallel_tool_calls:
+            output_text = "\n\n".join(
+                [tool_output.content for tool_output in tool_outputs]
+            )
+            return AgentChatResponse(response=output_text, sources=tool_outputs)
+        else:
+            if len(tool_outputs) > 1:
+                raise ValueError("Invalid")
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-mistralai/llama_index/llms/mistralai/base.py:440]: Same here

**Review by logan-markewich — APPROVED** [review]
Lgtm

**Pull request merged**